hal/targets/hal/TARGET_STM/TARGET_STM32F4/storage_driver.c

```diff
@@ -0,0 +1,570 @@
+    
+    /* look for sector to erase */
+    int8_t base_sector = -1;
+    uint8_t nb_sectors = 0;
+    uint8_t sector = STORAGE_START_SECTOR;
+    get_block(addr, &block);
+    tmp_addr = addr;
+    while (tmp_addr < (addr + size)) {
+        for (uint8_t i = 0; i < block.size / block.attributes.erase_unit; i++) {
+            uint32_t sector_addr = i * block.attributes.erase_unit + block.addr;
+            if (sector_addr >= addr && sector_addr + block.attributes.erase_unit <= addr + size) {
+                if (base_sector == -1) {
+                    base_sector = sector;
+                }
+                nb_sectors += 1;
+            }
+            sector += 1;
+        }
+        /* move on to the following block */
+        if (get_next_block(&block, &block) != ARM_DRIVER_OK) {
+            break;
+        }
+        tmp_addr = block.addr;
+    }
+        
+    FLASH_EraseInitTypeDef EraseInitStruct;
+    uint32_t sector_fail = 0;
+    
+    EraseInitStruct.TypeErase     = FLASH_TYPEERASE_SECTORS;
+    EraseInitStruct.VoltageRange  = FLASH_VOLTAGE_RANGE_3;
+    EraseInitStruct.Sector        = base_sector;
+    EraseInitStruct.NbSectors     = nb_sectors;
+    
+    if (capabilities.asynchronous_ops == 0) {
+        /* Synchronous erasing */
+        HAL_FLASH_Unlock();
+        if (HAL_FLASHEx_Erase(&EraseInitStruct, &sector_fail) != HAL_OK) {
+            HAL_FLASH_Lock();
+            return ARM_STORAGE_ERROR_NOT_ERASABLE;
+        }
+        HAL_FLASH_Lock();
+        return size;
+    } else {
+        /* Asynchronous erasing */
+        context->size = size;
+        HAL_FLASH_Unlock();
+        HAL_FLASHEx_Erase_IT(&EraseInitStruct);
+        return ARM_DRIVER_OK;
+    }
+}
+
+static int32_t erase_all(void)
+{
+    return ARM_DRIVER_ERROR_UNSUPPORTED;
+}
+
+static ARM_STORAGE_STATUS get_status(void)
+{
+    struct driver_context_s *context = &driver_context;
+    ARM_STORAGE_STATUS status = {
+        .busy  = 0,
+        .error = 0,
+    };
+    if (!context->initialized) {
+        status.error = 1;
+        return status;
+    }
+    if (pFlash.ProcedureOnGoing != FLASH_PROC_NONE) {
+        status.busy = 1;
+    } else if (pFlash.ErrorCode != HAL_FLASH_ERROR_NONE) {
+        status.error = 1;
+    }
+    return status;
+}
+
+static int32_t get_info(ARM_STORAGE_INFO *info)
+{
+    memcpy(info, &informations, sizeof(ARM_STORAGE_INFO));
+    return ARM_DRIVER_OK;
+}
+
+static uint32_t resolve_address(uint64_t addr)
+{
+    return (uint32_t)addr;
+}
+
+static int32_t get_next_block(const ARM_STORAGE_BLOCK *prev_block, ARM_STORAGE_BLOCK *next_block)
+{  
+    if (prev_block == NULL) {
+        /* fetching the first block (instead of next) */
+        if (next_block) {
+            const ARM_STORAGE_BLOCK *iter = &block_table[0];
+            uint8_t sector = 0;
+            for (size_t index = 0; index < block_table_size; ++index, ++iter) {
+                if (sector < STORAGE_START_SECTOR) {
```

Comment:
I'm not sure if STORAGE_START_SECTOR is the correct abstraction to use in this case. I've [posted a question](https://github.com/ARMmbed/mbed-os/pull/2556#pullrequestreview-2105272) to solicit input on this. It may be that STORAGE_START_SECTOR should remain as 0, and therefore implicit, and we use the storage-volume-manager to arbitrate access to partitions within a Storage driver.
I'm hoping @simonqhughes, @marcuschangarm and @sg- would be able to provide input and drive a consensus.